src/backtrace.c

```diff
@@ -0,0 +1,680 @@
+                        return;
+                    }
+                }
+                // FALLTHROUGH!
+            case BT_FUNCTION:
+                if (func.fp_offset)
+                    fp = *(uint32_t**)((uint32_t)sp + func.fp_offset);
+                ra = *(uint32_t**)((uint32_t)sp + func.ra_offset) - 2;
+                sp = (uint32_t*)((uint32_t)sp + func.stack_size);
+                exception_ra = NULL;
+                func_start = 0;
+                break;
+            case BT_EXCEPTION: {
+                // Exception frame. We must return back to EPC, but let's keep the
+                // RA value. If the interrupted function is a leaf function, we
+                // will need it to further walk back.
+                // Notice that FP is a callee-saved register so we don't need to
+                // recover it from the exception frame (also, it isn't saved there
+                // during interrupts).
+                exception_ra = *(uint32_t**)((uint32_t)sp + func.ra_offset);
+
+                // Read EPC from exception frame and adjust it with CAUSE BD bit
+                ra = *(uint32_t**)((uint32_t)sp + offsetof(reg_block_t, epc) + 32);
+                uint32_t cause = *(uint32_t*)((uint32_t)sp + offsetof(reg_block_t, cr) + 32);
+                if (cause & C0_CAUSE_BD) ra++;
+
+                sp = (uint32_t*)((uint32_t)sp + func.stack_size);
+
+                // Special case: if the exception is due to an invalid EPC
+                // (eg: a null function pointer call), we can rely on RA to get
+                // back to the caller. This assumes that we got there via a function call
+                // rather than a raw jump, but that's a reasonable assumption. It's anyway
+                // the best we can do.
+                if (C0_GET_CAUSE_EXC_CODE(cause) == EXCEPTION_CODE_TLB_LOAD_I_MISS &&
+                    !is_valid_address((uint32_t)ra)) {
+                    
+                    // Store the invalid address in the backtrace, so that it will appear in dumps.
+                    // This makes it easier for the user to understand the reason for the exception.
+                    cb(arg, ra);
+                    #if BACKTRACE_DEBUG
+                    debugf("backtrace: %s, ra=%p, sp=%p, fp=%p ra_offset=%d, fp_offset=%d, stack_size=%d\n", 
+                        "BT_INVALID", ra, sp, fp, func.ra_offset, func.fp_offset, func.stack_size);
+                    #endif
+                    
+                    ra = exception_ra - 2;
+
+                    // The function that jumped into an invalid PC was not interrupted by the exception: it
+                    // is a regular function
+                    // call now.
+                    exception_ra = NULL;
+                    break;
+                }
+
+                // The next frame might be a leaf function, for which we will not be able
+                // to find a stack frame. It is useful to try finding the function start.
```

Comment:
I didn't exactly get why this is useful only when we are in an exception frame? Considering it is only being used in tandem with `exception_ra`, maybe we won't need it? This part really confused me. I really feel like the logic around the two should be simpler, but I don't know how exactly they relate.